### PR TITLE
feature/554_batches_within_transactions

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #554 
* 100% unit tests passed:
```
Tests run: 59, Failures: 0, Errors: 0, Skipped: 0
```
* (unofficial) e2e tests passed:
```
-----------------------
OrionHDFSSink, json-row
-----------------------
time=2015-10-16T10:45:05.448CEST | lvl=INFO | trans=1444985093-152-0000000099 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[802] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data ({"recvTimeTs":"1444985101","recvTime":"2015-10-16T08:45:01.105Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"0","attrMd":[]}
{"recvTimeTs":"1444985101","recvTime":"2015-10-16T08:45:01.350Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"1","attrMd":[]}
{"recvTimeTs":"1444985101","recvTime":"2015-10-16T08:45:01.388Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"2","attrMd":[]}
...
{"recvTimeTs":"1444985105","recvTime":"2015-10-16T08:45:05.345Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"97","attrMd":[]}
{"recvTimeTs":"1444985105","recvTime":"2015-10-16T08:45:05.386Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"98","attrMd":[]}
{"recvTimeTs":"1444985105","recvTime":"2015-10-16T08:45:05.427Z","entityId":"Room1","entityType":"Room","attrName":"temperature","attrType":"centigrade","attrValue":"99","attrMd":[]})
...
hive> select * from frb_def_serv_deleteme_1_def_serv_path_deleteme_1_room1_room_row;
OK
1444985101	2015-10-16T08:45:01.105Z	Room1	Room	temperature	centigrade	0	[]
1444985101	2015-10-16T08:45:01.350Z	Room1	Room	temperature	centigrade	1	[]
1444985101	2015-10-16T08:45:01.388Z	Room1	Room	temperature	centigrade	2	[]
...
1444985105	2015-10-16T08:45:05.345Z	Room1	Room	temperature	centigrade	97	[]
1444985105	2015-10-16T08:45:05.386Z	Room1	Room	temperature	centigrade	98	[]
1444985105	2015-10-16T08:45:05.427Z	Room1	Room	temperature	centigrade	99	[]
--------------------------
OrionHDFSSink, json-column
--------------------------
time=2015-10-16T11:01:03.834CEST | lvl=INFO | trans=1444986053-614-0000000099 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[802] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data ({"recvTime":"2015-10-16T09:00:59.665Z", "temperature":"0", "temperature_md":[]}
{"recvTime":"2015-10-16T09:00:59.759Z", "temperature":"1", "temperature_md":[]}
{"recvTime":"2015-10-16T09:00:59.798Z", "temperature":"2", "temperature_md":[]}
...
{"recvTime":"2015-10-16T09:01:03.727Z", "temperature":"97", "temperature_md":[]}
{"recvTime":"2015-10-16T09:01:03.765Z", "temperature":"98", "temperature_md":[]}
{"recvTime":"2015-10-16T09:01:03.805Z", "temperature":"99", "temperature_md":[]})
...
hive> select * from frb_def_serv_deleteme_1_def_serv_path_deleteme_1_room1_room_column;
OK
2015-10-16T09:00:59.665Z	0	[]
2015-10-16T09:00:59.759Z	1	[]
2015-10-16T09:00:59.798Z	2	[]
...
2015-10-16T09:01:03.727Z	97	[]
2015-10-16T09:01:03.765Z	98	[]
2015-10-16T09:01:03.805Z	99	[]
----------------------
OrionHDFSSink, csv-row
----------------------
time=2015-10-16T11:08:44.099CEST | lvl=INFO | trans=1444986514-772-0000000099 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[802] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data (1444986519,2015-10-16T09:08:39.535Z,Room1,Room,temperature,centigrade,0,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986519,2015-10-16T09:08:39.701Z,Room1,Room,temperature,centigrade,1,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986519,2015-10-16T09:08:39.739Z,Room1,Room,temperature,centigrade,2,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
1444986523,2015-10-16T09:08:43.723Z,Room1,Room,temperature,centigrade,97,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986523,2015-10-16T09:08:43.865Z,Room1,Room,temperature,centigrade,98,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986523,2015-10-16T09:08:43.982Z,Room1,Room,temperature,centigrade,99,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt)
...
time=2015-10-16T12:24:49.622CEST | lvl=INFO | trans=1444991078-683-0000000099 | function=persistMDAggregations | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[816] : [hdfs-sink] Persisting metadata at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt), Data (1444991083121,resolution,integer,0
1444991083216,resolution,integer,.50
1444991083255,resolution,integer,1.00
...
1444991087907,resolution,integer,48.50
1444991087965,resolution,integer,49.00
1444991088010,resolution,integer,49.50)
...
hive> select * from frb_def_serv_deleteme_1_def_serv_path_deleteme_1_room1_room_row;   
OK
1444986519	2015-10-16T09:08:39.535Z	Room1	Room	temperature	centigrade	0	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986519	2015-10-16T09:08:39.701Z	Room1	Room	temperature	centigrade	1	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986519	2015-10-16T09:08:39.739Z	Room1	Room	temperature	centigrade	2	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
1444986523	2015-10-16T09:08:43.723Z	Room1	Room	temperature	centigrade	97	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986523	2015-10-16T09:08:43.865Z	Room1	Room	temperature	centigrade	98	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444986523	2015-10-16T09:08:43.982Z	Room1	Room	temperature	centigrade	99	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
$ hadoop fs -cat /user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444991083121,resolution,integer,0
1444991083216,resolution,integer,.50
1444991083255,resolution,integer,1.00
...
1444991087907,resolution,integer,48.50
1444991087965,resolution,integer,49.00
1444991088010,resolution,integer,49.50
-------------------------
OrionHDFSSink, csv-column
-------------------------
time=2015-10-16T12:28:43.368CEST | lvl=INFO | trans=1444991314-950-0000000099 | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[798] : [hdfs-sink] Persisting data at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room/room1_room.txt), Data (2015-10-16T10:28:38.981Z,0,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:39.111Z,1,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:39.153Z,2,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
2015-10-16T10:28:43.255Z,97,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:43.294Z,98,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:43.332Z,99,hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
time=2015-10-16T12:28:44.671CEST | lvl=INFO | trans=1444991314-950-0000000099 | function=persistMDAggregations | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionHDFSSink[816] : [hdfs-sink] Persisting metadata at OrionHDFSSink. HDFS file (def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt), Data (1444991318981,resolution,integer,0
1444991319111,resolution,integer,.50
1444991319153,resolution,integer,1.00
...
1444991323255,resolution,integer,48.50
1444991323294,resolution,integer,49.00
1444991323332,resolution,integer,49.50)
...
hive> select * from frb_def_serv_deleteme_1_def_serv_path_deleteme_1_room1_room_column;
OK
2015-10-16T10:28:38.981Z	0	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:39.111Z	1	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:39.153Z	2	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
2015-10-16T10:28:43.255Z	97	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:43.294Z	98	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
2015-10-16T10:28:43.332Z	99	hdfs:///user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
...
$ hadoop fs -cat /user/frb/def_serv_deleteme_1/def_serv_path_deleteme_1/room1_room_temperature_centigrade/room1_room_temperature_centigrade.txt
1444991318981,resolution,integer,0
1444991319111,resolution,integer,.50
1444991319153,resolution,integer,1.00
...
1444991323255,resolution,integer,48.50
1444991323294,resolution,integer,49.00
1444991323332,resolution,integer,49.50
```
* Assignee @fgalan

### CODE REVIEW

**Comment by frbattid** [comment]
Tech debt issues raised from this PR:

* https://github.com/telefonicaid/fiware-cygnus/issues/562
* https://github.com/telefonicaid/fiware-cygnus/issues/563
* https://github.com/telefonicaid/fiware-cygnus/issues/564

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1 +1,2 @@
 - [FEATURE] Add a binary implementation of the HDFS backend (#537)
+- [FEATURE] Add support to batch processing within a transaction (#554)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: ...for the HDFS sink

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java [hunk 2]
```diff
@@ -255,16 +322,70 @@ private void persist(Event event) throws Exception {
...
+            
+            CygnusEvent cygnusEvent = new CygnusEvent(
+                    recvTimeTs, service, groupedServicePaths[i], destination,
+                    notification.getContextResponses().get(i).getContextElement());
+            list.add(cygnusEvent);
+        } // for
+    } // accumulateInBatch
+    
+    private void doBatchRollback() {
+        // TBD: tech debt
```
Comment (fgalan) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: Include issue number

File: doc/design/OrionHDFSSink.md [hunk 3]
```diff
@@ -203,21 +221,10 @@ A configuration example could be:
...
-
-There exists an [issue](https://github.com/telefonicaid/fiware-cosmos/issues/111) about adding OAuth2 support to the Hadoop RPC mechanism, in the context of the [`fiware-cosmos`](https://github.com/telefonicaid/fiware-cosmos) project.
-
-[Top](#top)
-
 ##<a name="section3"></a>Use cases
-Use `OrionHDFSSink` if you are looking for a JSON or CSV-based document storage growing in the mid-long term.
+Use `OrionHDFSSink` if you are looking for a JSON or CSV-based document storage growing in the mid-long-term in estimated sizes of terabytes for future trending discovery, along the time persistent patterns of behaviour and so on.
+
+For a short-term historic, those required by dashboards and charting user interfaces, other backends are more suited such as MongoDB or MySQL (Cygnus provides sinks for them, as well).
```
Comment (fgalan) [inline doc/design/OrionHDFSSink.md]: Maybe FIWARE Comet should be refered from Cygnus documentation? Not only in this specific point, but in every pooint in which short-term historic or STH feature is mentioned...

File: doc/design/OrionHDFSSink.md [hunk 4]
```diff
@@ -225,13 +232,13 @@ Use `OrionHDFSSink` if you are looking for a JSON or CSV-based document storage
 ###<a name="section4.1"></a>`OrionHDFSSink` class
 As any other NGSI-like sink, `OrionHDFSSink` extends the base `OrionSink`. The methods that are extended are:
```
Comment (fgalan) [inline doc/design/OrionHDFSSink.md]: It should be avoided to include too much deep developer-oriented information in users manuals, e.g. this part showing method signature is a big "hard". I'd suggest to analyze this globally and do the proper documentation refator if needed.

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/b13eedc0daf7fb440f3d3decb43b596a54cb9f1f

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java (hunk 2 above)
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/b13eedc0daf7fb440f3d3decb43b596a54cb9f1f

File: doc/design/OrionHDFSSink.md (hunk 3 above)
Comment (frbattid) [inline doc/design/OrionHDFSSink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/b13eedc0daf7fb440f3d3decb43b596a54cb9f1f

File: doc/design/OrionHDFSSink.md (hunk 4 above)
Comment (frbattid) [inline doc/design/OrionHDFSSink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/commit/b13eedc0daf7fb440f3d3decb43b596a54cb9f1f

**Comment by fgalan** [comment]
LGTM

**Pull request merged**